src/runtime/rpc/network.cpp

```diff
@@ -478,29 +457,24 @@ bool rpc_session::on_recv_message(message_ex *msg, int delay_ms)
     return true;
 }
 
-void rpc_session::start_negotiation()
+void rpc_session::set_negotiation_succeed()
 {
-    if (security::FLAGS_enable_auth) {
-        // set the negotiation state if it's a client rpc_session
-        if (is_client()) {
-            set_negotiation();
-        }
+    utils::auto_lock<utils::ex_lock_nr> l(_lock);
+    negotiation_succeed = true;
 
-        auth_negotiation();
-    } else {
-        // set negotiation success if auth is disabled
-        on_success();
-    }
+    // todo(zlw): resend pending messages when negotiation is succeed
 }
 
-void rpc_session::auth_negotiation()
+bool rpc_session::is_negotiation_succeed()
 {
-    _negotiation = security::create_negotiation(is_client(), this);
-    _negotiation->start();
+    if (negotiation_succeed) {
+        return negotiation_succeed;
```

Comment:
Why these sentense out of lock?

Reply:
this is kind of double check. the first one didn't lock the `_lock`.
Because `negotiation_succeed` only transfered from `false` to `true`. So if it is true now, it will not change in the later. but if it is `false` now, maybe it will change soon. So we should use lock to protect it.
Why I use double check here is to be efficient

Reply:
OK, better to add some comments here

Reply:
I removed the double check and the lock operation in `is_negotiation_succeed`. Because I think it is ok to get the value of `negotiation_succeed` without lock. 
The double check will be used in `set_negotiation_succeed` in the later pull request.

Reply:
Is `on_recv_message` in the same lock?

Reply:
on_recv_message will not lock any locks.  It only calls `is_negotiation_succeed` to judge whether the negotiation is succeed

Reply:
Another better way to implement it is to change the type of `_lock` from `utils::ex_lock_nr` to `zrwlock_nr`. And in `is_negotiation_succeed` we use `zauto_read_lock l(_lock)` to lock read. And in other cases we should use `zauto_write_lock l(_lock)`